src/domain/task-system/src/aggregates/task.rs

```diff
@@ -0,0 +1,276 @@
+    last_synced_event: Option<EventID>,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+impl Task {
+    pub fn new(task_id: TaskID, logical_plan: LogicalPlan) -> Self {
+        let genesis = TaskCreated {
+            event_time: Utc::now(),
+            task_id,
+            logical_plan,
+        };
+
+        Self {
+            state: TaskState {
+                task_id,
+                status: TaskStatus::Queued,
+                cancellation_requested: false,
+                logical_plan: genesis.logical_plan.clone(),
+                created_at: genesis.event_time.clone(),
+                ran_at: None,
+                cancellation_requested_at: None,
+                finished_at: None,
+            },
+            pending_events: vec![genesis.into()],
+            last_synced_event: None,
+        }
+    }
+
+    /// Transition task to a `Running` state
+    pub fn run(&mut self) -> Result<(), IllegalSequenceError<Self>> {
+        self.apply(TaskRunning {
+            event_time: Utc::now(),
+            task_id: self.task_id,
+        })
+    }
+
+    /// Task is queued or running and cancellation was not already requested
+    pub fn can_cancel(&mut self) -> bool {
+        match self.status {
+            TaskStatus::Queued if !self.cancellation_requested => true,
+            TaskStatus::Running if !self.cancellation_requested => true,
+            _ => false,
+        }
+    }
+
+    /// Set cancellation flag (if not already set)
+    pub fn cancel(&mut self) -> Result<(), IllegalSequenceError<Self>> {
+        if self.cancellation_requested {
+            return Ok(());
+        }
+
+        self.apply(TaskCancelled {
+            event_time: Utc::now(),
+            task_id: self.task_id,
+        })
+    }
+
+    /// Transition task to a `Finished` state with the specified outcome
+    pub fn finish(&mut self, outcome: TaskOutcome) -> Result<(), IllegalSequenceError<Self>> {
+        self.apply(TaskFinished {
+            event_time: Utc::now(),
+            task_id: self.task_id,
+            outcome,
+        })
+    }
+
+    fn apply(&mut self, event: impl Into<TaskEvent>) -> Result<(), IllegalSequenceError<Self>> {
+        let event = event.into();
+        self.update_state(event.clone())?;
+        self.pending_events.push(event);
+        Ok(())
+    }
+
+    fn update_state(&mut self, event: TaskEvent) -> Result<(), IllegalSequenceError<Self>> {
+        assert_eq!(self.state.task_id, event.task_id());
+
+        // Check if state transition is legal
+        match (self.state.status, &event) {
+            (TaskStatus::Queued, TaskEvent::Running(_)) => {}
+            (TaskStatus::Queued | TaskStatus::Running, TaskEvent::Cancelled(_)) => {}
+            (TaskStatus::Queued | TaskStatus::Running, TaskEvent::Finished(_)) => {}
+            (_, _) => return Err(IllegalSequenceError::new(self.state.clone(), event).into()),
+        }
```

Comment:
Are Queued->Queued, Running->Running, Finished->Finished, Cancelled->Cancelled possible?
Probably these should be just ignored to satisfy idempotence property.

Reply:
`Queued->Queued` is impossible as this is the initial state.

For the rest ... not sure. I think I would rather be strict and not allow them in the aggregate.

There is a separate question of conflict resolution, e.g. if two users concurrently cancel the same task should we silently deduplicate the event on save? My instincts is that this adds too much complexity, and I would rather have optimistic concurrency errors and retry the transaction a few times than have an overly smart deduplication logic at the event store level.